src/lib/homestore_backend/replication_state_machine.cpp

```diff
@@ -159,30 +164,139 @@ ReplicationStateMachine::create_snapshot(std::shared_ptr< homestore::snapshot_co
+    return true;
 }
 
-std::shared_ptr< homestore::snapshot_context > ReplicationStateMachine::last_snapshot() {
-    LOGE("last_snapshot not implemented");
-    return nullptr;
-}
+std::shared_ptr< homestore::snapshot_context > ReplicationStateMachine::last_snapshot() { return m_snapshot_context; }
 
 int ReplicationStateMachine::read_snapshot_data(std::shared_ptr< homestore::snapshot_context > context,
                                                 std::shared_ptr< homestore::snapshot_data > snp_data) {
-    LOGE("read_snapshot_data not implemented");
-    return -1;
+    HSHomeObject::PGBlobIterator* pg_iter = nullptr;
+    auto s = dynamic_pointer_cast< homestore::nuraft_snapshot_context >(context)->nuraft_snapshot();
+
+    if (snp_data->user_ctx == nullptr) {
+        // Create the pg blob iterator for the first time.
+        pg_iter = new HSHomeObject::PGBlobIterator(*home_object_, repl_dev()->group_id());
+        snp_data->user_ctx = (void*)pg_iter;
+    } else {
+        pg_iter = r_cast< HSHomeObject::PGBlobIterator* >(snp_data->user_ctx);
+    }
+
+    // Nuraft uses obj_id as a way to track the state of the snapshot read and write.
+    // Nuraft starts with obj_id == 0 as first message always, leader send all the shards and
+    // PG metadata as response. Follower responds with next obj_id it expects. obj_id's are
+    // encoded in the form of obj_id = shard_seq_num(6 bytes) | batch_number(2 bytes)
+    // Leader starts with shard sequence number 1 and read upto maximum size of data
+    // and send to follower in a batch. Once all blob's are send in a shard,
+    // leader notifies the follower by setting end_of_batch in the payload. Follower
+    // moves to the next shard by incrementing shard_seq_num and reset batch number to 0.
+    // Batch number is used to identify which batch in the current shard sequence number.
+    // We use pg blob iterator to go over all the blobs in all the shards in that PG.
+    // Once all the blobs are finished sending, end_of_scan will be true.
+    int64_t obj_id = snp_data->offset;
+    uint64_t shard_seq_num = obj_id >> 16;
+    uint64_t batch_number = obj_id & 0xFFFF;
+    auto log_str = fmt::format("group={} term={} lsn={} shard_seq={} batch_num={} size={}",
+                               boost::uuids::to_string(repl_dev()->group_id()), s->get_last_log_term(),
+                               s->get_last_log_idx(), shard_seq_num, batch_number, snp_data->blob.size());
+    if (obj_id == 0) {
+        // obj_id = 0 means its the first message and we send the pg and its shards metadata.
+        cur_snapshot_batch_num = 0;
+        pg_iter->create_pg_shard_snapshot_data(snp_data->blob);
+        RELEASE_ASSERT(snp_data->blob.size() > 0, "Empty metadata snapshot data");
+        LOGD("Read snapshot data first message {}", log_str);
+        return 0;
+    }
+
+    if (shard_seq_num != pg_iter->cur_shard_seq_num_ || batch_number != cur_snapshot_batch_num) {
```

Comment:
the implementation right now is not friendly for retry.   writer side cannot re-try on shard level if it failed to apply some batch.

for example,   if {shard x, batch 1} is the last batch of shard x,  the reader(leader) side will set 
```
cur_shard_seq_num_ = x+1
cur_snapshot_batch_num = 0
```
writer (follower) send back {x, 1} will result a failure.

IIRC we agreed in the design if failure happens in {x, y} , we rewind to {x, 0}

Reply:
my general suggestion is changing pg_iter to shard_iter, and let the writer controls the pace of which shard it is requesting. 

we can maintain a bitmap `is_shard_completed[shardID]` to track if a shard are fully transferred.  and based on that we know we are fully done (set the `is_last_obj = true`).

some enhancement can be :
1.  If writer side can confirm a shard done by sending next_objid = {shardid,  0} if our batch started with 1.
2.  if writer side can tell reader it has all shards (based on the list it got from first step) ,  leader can check if it does reach to the end,  if not , we can better error out.

Reply:
Yeah agree. we need to support for retry for errors or if it crashed and came back. There is a chance it can go back.  If we see a difference in shard_sequence or batch number and is less than current, then we know follower is re-requesting the data and we can accept it as our current. We can continue from what the follower has provided.

Reply:
yeah just do not trust batch number... the batch will change by IO, 
for example, with batchsize = 3
```
B1{lsn 1}, B2{lsn 2}, B3{{lsn 3}, B4 {lsn 4}, B5{lsn 5}, B6{lsn 6}
```
will become
```
batch 1:
B1-B3
batch 2:
B4-B6
```

later on if delete comes in for B1-B3
it will be 
```
B1{TombStone 11}, B2{TombStone 12}, B3{{TombStone 13}, B4 {lsn 4}, B5{lsn 5}, B6{lsn 6}
```
so we only have  
```
batch 1:
B4-B6
```